File: apps/worker/adapters/qatar_award.py
# apps/worker/adapters/qatar_award.py
#
# Two strategies (tried in order):
#   1. Seats.aero API  — if SEATS_AERO_API_KEY is set
#   2. Qatar Privilege Club direct Playwright scrape

import os
import re
import time
import random
import requests
from playwright.sync_api import sync_playwright, Browser, BrowserContext


def _sleep(ms: int, jitter: int = 0):
    time.sleep((ms + random.randint(0, jitter)) / 1000)


# ── Seats.aero ──────────────────────────────────────────────────────────────

def _scrape_seats_aero(monitor: dict, api_key: str) -> list[dict]:
    year, month = monitor["month"].split("-")
    start_date = f"{year}-{month}-01"
    end_date = f"{year}-{month}-28"
    destination = monitor["destination"]

    url = (
        "https://api.seats.aero/partnerapi/availability"
        f"?origin_airport=YYZ&destination_airport={destination}"
        f"&start_date={start_date}&end_date={end_date}"
        "&cabin=business&source=qr"
    )
    headers = {
        "Partner-Authorization": api_key,
        "Content-Type": "application/json",
    }

    try:
        resp = requests.get(url, headers=headers, timeout=20)
        if not resp.ok:
            print(f"  [SeatsAero] HTTP {resp.status_code}")
            return []

        trips = resp.json().get("data", [])
        results = []
        for t in trips:
            if not (t.get("JAvailable") or t.get("YAvailable")):
                continue
            results.append({
                "provider": "Seats.aero",
                "monitor_id": monitor["id"],
                "kind": "award",
                "origin": "YYZ",
                "destination": destination,
                "departure_date": t.get("Date", f"{year}-{month}-01"),
                "points_cost": t.get("JMileageCost") or t.get("YMileageCost") or 0,
                "cash_surcharge": t.get("JTaxes") or t.get("YTaxes") or 0,
                "currency": "USD",
                "cabin": "business",
                "airline": "Qatar Airways",
                "stops": 1,
                "duration": "~18h",
                "booking_url": (
                    "https://www.qatarairways.com/en/privilege-club/"
                    "redeem/flight-rewards.html"
                ),
                "is_award": True,
            })

        print(f"  [SeatsAero] {len(results)} available dates for {monitor['id']}")
        return results

    except Exception as e:
        print(f"  [SeatsAero] Error: {e}")
        return []


# ── Qatar Privilege Club direct scrape ──────────────────────────────────────

def _scrape_qatar_direct(monitor: dict) -> list[dict]:
    year, month = monitor["month"].split("-")
    destination = monitor["destination"]
    results = []

    booking_url = (
        "https://www.qatarairways.com/en/privilege-club/redeem/flight-rewards.html"
        f"?widget=QR&searchType=F&addTaxToMiles=on&bookingClass=J"
        f"&tripType=O&fromStation=YYZ&toStation={destination}"
        f"&departingHidden={year}-{month}-01"
        "&returnHidden=&numOfAdults=1&numOfChildren=0&numOfInfants=0"
    )

    with sync_playwright() as p:
        browser: Browser = p.chromium.launch(
            headless=True,
            args=["--no-sandbox", "--disable-setuid-sandbox"],
        )
        context: BrowserContext = browser.new_context(
            user_agent=(
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/122.0.0.0 Safari/537.36"
            ),
            locale="en-GB",
        )
        context.add_init_script(
            "Object.defineProperty(navigator, 'webdriver', {get: () => false})"
        )
        page = context.new_page()

        try:
            print(f"  [QatarDirect] Loading {destination} {monitor['month']}")
            page.goto(booking_url, wait_until="domcontentloaded", timeout=30000)
            _sleep(4000, 2000)

            # Try clicking search button
            for selector in ['[id*="btnSearch"]', 'button[type="submit"]']:
                try:
                    btn = page.query_selector(selector)
                    if btn and btn.is_visible():
                        btn.click()
                        _sleep(5000, 2000)
                        break
                except Exception:
                    pass

            # Find available date cells
            available = page.query_selector_all(
                ".calendar-cell.available, "
                "[class*='available'][class*='date'], "
                "[data-available='true']"
            )
            print(f"  [QatarDirect] Found {len(available)} available dates")

            for i, cell in enumerate(available[:20]):
                try:
                    date_attr = cell.get_attribute("data-date")
                    text = cell.inner_text()
                    miles_match = re.search(r'([\d,]+)\s*miles?', text, re.I)
                    miles = int(miles_match.group(1).replace(",", "")) if miles_match else None

                    dep_date = date_attr or f"{year}-{month}-{str(i+1).zfill(2)}"

                    if miles:
                        results.append({
                            "provider": "Qatar Airways (Direct)",
                            "monitor_id": monitor["id"],
                            "kind": "award",
                            "origin": "YYZ",
                            "destination": destination,
                            "departure_date": dep_date,
                            "points_cost": miles,
                            "cash_surcharge": 250,
                            "currency": "USD",
                            "cabin": monitor.get("cabin", "business"),
                            "airline": "Qatar Airways",
                            "stops": 1,
                            "duration": "~18h",
                            "booking_url": booking_url,
                            "is_award": True,
                        })
                except Exception:
                    continue

            # If cells exist but no miles parsed, still flag availability
            if not results and available:
                results.append({
                    "provider": "Qatar Airways (Direct)",
                    "monitor_id": monitor["id"],
                    "kind": "award",
                    "origin": "YYZ",
                    "destination": destination,
                    "departure_date": f"{year}-{month}",
                    "points_cost": None,
                    "currency": "USD",
                    "cabin": monitor.get("cabin", "business"),
                    "airline": "Qatar Airways",
                    "stops": 1,
                    "duration": "~18h",
                    "booking_url": booking_url,
                    "is_award": True,
                })

        except Exception as e:
            print(f"  [QatarDirect] Error: {e}")
        finally:
            page.close()
            browser.close()

    return results


# ── Public entry point ───────────────────────────────────────────────────────

def scrape_award_month(monitor: dict) -> list[dict]:
    seats_aero_key = os.environ.get("SEATS_AERO_API_KEY", "")
    if seats_aero_key:
        results = _scrape_seats_aero(monitor, seats_aero_key)
        if results:
            return results
    return _scrape_qatar_direct(monitor)

File: apps/worker/src/adapters/qatar-award.ts
// apps/worker/src/adapters/qatar-award.ts
// Scrapes Qatar Airways Privilege Club for award availability
// Primary: Qatar Airways website | Fallback: seats.aero API

import axios from "axios";
import { chromium, Browser, BrowserContext } from "playwright";
import { CashMonitor, AwardMonitor, FlightResult } from "@flight-tracker/shared";
import { ScrapeAdapter, buildFingerprint, sleep, retryWithBackoff } from "./base";

const PROXY_CONFIG = process.env.PROXY_ENDPOINT
  ? {
      server: process.env.PROXY_ENDPOINT,
      username: process.env.PROXY_USERNAME,
      password: process.env.PROXY_PASSWORD,
    }
  : undefined;

export class QatarAwardAdapter implements ScrapeAdapter {
  name = "Qatar Airways (Award)";
  priority = 1;

  private browser: Browser | null = null;
  private context: BrowserContext | null = null;

  async isHealthy(): Promise<boolean> {
    return true;
  }

  async scrapeWindow(_monitor: CashMonitor): Promise<FlightResult[]> {
    return []; // Cash fares not supported
  }

  async scrapeMonth(monitor: AwardMonitor): Promise<FlightResult[]> {
    // Try seats.aero first (faster, more reliable) if API key present
    if (process.env.SEATS_AERO_API_KEY) {
      try {
        const results = await retryWithBackoff(
          () => this.scrapeSeatsAero(monitor),
          2,
          3000
        );
        if (results.length > 0) return results;
      } catch (err: any) {
        console.warn(`[Qatar-Award] seats.aero failed: ${err.message}`);
      }
    }

    // Fallback to Playwright scraping of Qatar Privilege Club
    return retryWithBackoff(
      () => this.scrapeQatarWebsite(monitor),
      3,
      8000
    );
  }

  // ─── seats.aero API ────────────────────────────────────────────────

  private async scrapeSeatsAero(monitor: AwardMonitor): Promise<FlightResult[]> {
    const [year, monthNum] = monitor.month.split("-");
    const startDate = `${year}-${monthNum}-01`;
    const endDate = `${year}-${monthNum}-${this.daysInMonth(parseInt(year), parseInt(monthNum))}`;

    const response = await axios.get("https://seats.aero/partnerapi/availability", {
      headers: {
        "Partner-Authorization": process.env.SEATS_AERO_API_KEY!,
        "Content-Type": "application/json",
      },
      params: {
        origin_airport: monitor.origin,
        destination_airport: monitor.destination,
        start_date: startDate,
        end_date: endDate,
        cabin: this.mapCabinToSeatsAero(monitor.cabin),
      },
      timeout: 15000,
    });

    const data = response.data?.data ?? [];
    const results: FlightResult[] = [];

    for (const slot of data) {
      if (!slot.available) continue;

      const result: FlightResult = {
        provider: "seats.aero",
        monitorId: monitor.id,
        kind: "award",
        origin: monitor.origin,
        destination: monitor.destination,
        departureDate: slot.date,
        pointsCost: slot.mileageCost ?? slot.points,
        cashSurcharge: slot.taxesFees,
        currency: "CAD",
        cabin: monitor.cabin ?? "business",
        airline: "Qatar Airways",
        stops: slot.stops ?? 1,
        duration: slot.duration ?? "Unknown",
        bookingUrl: `https://www.qatarairways.com/en-ca/privilege-club/redeem-avios.html`,
        scrapedAt: new Date().toISOString(),
      };
      (result as any).fingerprint = buildFingerprint(result);
      results.push(result);
    }

    return results;
  }

  // ─── Qatar Airways Website Scraper ────────────────────────────────

  private async scrapeQatarWebsite(monitor: AwardMonitor): Promise<FlightResult[]> {
    if (!this.browser) await this.initBrowser();

    const page = await this.context!.newPage();
    const results: FlightResult[] = [];

    try {
      // Navigate to Qatar Privilege Club search
      const url = `https://www.qatarairways.com/en-ca/privilege-club/redeem-avios.html`;
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
      await sleep(3000);

      // Fill in search form
      // Origin
      const originInput = await page.waitForSelector(
        'input[placeholder*="From"], input[name*="origin"], [data-testid*="origin"]',
        { timeout: 10000 }
      );
      await originInput?.fill(monitor.origin);
      await sleep(1000);
      await page.keyboard.press("ArrowDown");
      await page.keyboard.press("Enter");

      // Destination
      const destInput = await page.waitForSelector(
        'input[placeholder*="To"], input[name*="destination"], [data-testid*="destination"]',
        { timeout: 8000 }
      );
      await destInput?.fill(monitor.destination);
      await sleep(1000);
      await page.keyboard.press("ArrowDown");
      await page.keyboard.press("Enter");

      // Date - set to first day of the month
      const [year, monthNum] = monitor.month.split("-");
      const dateStr = `${year}-${monthNum}-01`;

      const dateInput = await page.waitForSelector(
        'input[type="date"], [data-testid*="date"]',
        { timeout: 8000 }
      );
      await dateInput?.fill(dateStr);
      await sleep(500);

      // Submit search
      const searchBtn = await page.waitForSelector(
        'button[type="submit"], [data-testid*="search"]',
        { timeout: 8000 }
      );
      await searchBtn?.click();
      await sleep(5000);

      // Parse results
      try {
        await page.waitForSelector(
          '[class*="flight-result"], [class*="award-result"], [data-testid*="flight"]',
          { timeout: 15000 }
        );

        const flights = await page.evaluate(() => {
          const cards = document.querySelectorAll(
            '[class*="flight-result"], [class*="award"], [data-testid*="flight"]'
          );
          return Array.from(cards).map((card) => {
            const pointsEl = card.querySelector('[class*="points"], [class*="miles"], [class*="avios"]');
            const points = parseInt(
              pointsEl?.textContent?.replace(/\D/g, "") ?? "0"
            );
            const dateEl = card.querySelector('[class*="date"], time');
            const dateText = dateEl?.textContent?.trim() ?? "";
            const durationEl = card.querySelector('[class*="duration"]');
            const duration = durationEl?.textContent?.trim() ?? "Unknown";
            return { points, dateText, duration };
          });
        });

        for (const f of flights) {
          if (!f.points) continue;
          const result: FlightResult = {
            provider: this.name,
            monitorId: monitor.id,
            kind: "award",
            origin: monitor.origin,
            destination: monitor.destination,
            departureDate: dateStr,
            pointsCost: f.points,
            currency: "CAD",
            cabin: monitor.cabin ?? "business",
            airline: "Qatar Airways",
            stops: 1,
            duration: f.duration,
            bookingUrl: url,
            scrapedAt: new Date().toISOString(),
          };
          (result as any).fingerprint = buildFingerprint(result);
          results.push(result);
        }
      } catch {
        // No results found for this month
        console.log(`[Qatar-Award] No award seats found for ${monitor.id}`);
      }
    } finally {
      await page.close();
    }

    return results;
  }

  private async initBrowser(): Promise<void> {
    this.browser = await chromium.launch({
      headless: true,
      args: ["--no-sandbox", "--disable-setuid-sandbox"],
    });
    this.context = await this.browser.newContext({
      proxy: PROXY_CONFIG,
      userAgent:
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
      viewport: { width: 1280, height: 900 },
      locale: "en-CA",
    });
    await this.context.addInitScript(() => {
      Object.defineProperty(navigator, "webdriver", { get: () => false });
    });
  }

  async close(): Promise<void> {
    await this.context?.close();
    await this.browser?.close();
    this.browser = null;
    this.context = null;
  }

  private mapCabinToSeatsAero(
    cabin?: string
  ): string {
    const map: Record<string, string> = {
      economy: "Y",
      premium_economy: "W",
      business: "J",
      first: "F",
    };
    return map[cabin ?? "business"] ?? "J";
  }

  private daysInMonth(year: number, month: number): string {
    return new Date(year, month, 0).getDate().toString().padStart(2, "0");
  }
}
